### PR TITLE
 `modified_files` and `modified_lines_in_file` on CommitMsg 

### LINKED ISSUE #726: `modified_files` and `modified_lines_in_file` on CommitMsg
I want to create a hook that makes sure that the commit has a longer body describing the changes in detail but only when more than 30 lines are changed across 3 or more files.

Would it be possible for `modified_files` to work the same way as it does in PreCommit hook?

EDIT: I think I would need `modified_lines_in_file` as well for this.

### PR DESCRIPTION
**Pull request by Quintasan**
As I promised – I did try my hand and I employed the dumbest approach that's almost not copy & paste*

## Big problem

It's 2 in the morning so I'm probably cognitively impaired enough to not be able to figure out why `rspec spec/overcommit/hook_context/commit_msg_spec.rb:636:705:764` are failing. 

All those fails have amending in common so I suspect this is due to calling `amendment?` not from the `Overcommit::HookContext::CommitMsg` hook context but from the `Overcommit::HookContext::PreCommit` context which I created using `Overcommit::HookContext::PreCommit.new(@config, @args, @input)` but I am unable to pinpoint why would it behave differently.

## Good points

1. Reuses a big chunk of `Overcommit::HookContext::PreCommit` so in the future you'd have to modify only that

## Bad points

1. `store_modified_times` is copypasta
2. `restore_modified_times` is copypasta
3. `modified_lines_in_file` is quite close to being copypasta
4. Test are copypasta – can be solved by converting into shared examples*

Fixing 1, 2, 3 would require some changes in the `PreCommit` hook context but I didn't want to throw a huge PR right away without asking first.

@sds Please let me know if this looks reasonable or I need another approach.

Closes #726

### CODE REVIEW

**Comment by sds** [comment]
Thanks for the PR, @Quintasan. Supportive of the general direction.

Looking over this quickly, I don't have an obvious explanation for the failing tests, but my guess would be we'd probably have an easier time extracting the shared logic into a module that we `include` in both `Overcommit::HookContext::CommitMsg` and `Overcommit::HookContext::PreCommit`.

I'm also OK having the tests be duplicated rather than using shared examples. There are often subtle differences between hooks and I don't think DRY tests in this context would be worth it, even if right now we expect the same behavior ultimately.

Appreciate you proposing this general direction. If we can extract the environment setup/cleanup logic into a module and get the tests passing that way, I'd be comfortable merging. By extracting into a module, perhaps we could finally (separate from this PR, of course) consider implementing a mechanism that doesn't use the stash (see #238).

**Comment by Quintasan** [comment]
Great, I hope to have something you can work with in the upcoming week

**Comment by Quintasan** [comment]
@sds It turns out I couldn't wait this much to work on this.

I wasn't sure where to put the extracted modules so I went with `Helpers` module but I think `SharedBehavior` would work better

WDYT?

**Review by sds — COMMENTED** [review]
Overall looks good. Some minor comments to address and then we'll happily merge. Thanks!

File: lib/overcommit/hook_context/helpers/stash_unstaged_changes.rb [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                "Unable to setup environment for #{hook_script_name} hook run:" \
+                "\nSTDOUT:#{result.stdout}\nSTDERR:#{result.stderr}"
+        end
+
+        @changes_stashed = `git stash list -1`.include?(stash_message)
+      end
+
+      # Restores the file modification times for all modified files to make it
+      # appear like they never changed.
+      def restore_modified_times
```
[the remark this answers is not in the document]
Comment (Quintasan) [reply lib/overcommit/hook_context/helpers/stash_unstaged_changes.rb:100]: @sds the intention here was to provide a way to reuse methods related to file modifications **without** that the workflow always will be 'we must hide unstaged changes for the hook context'.

File: Gemfile [hunk 2]
```diff
@@ -14,4 +14,7 @@ gem 'coveralls', '~> 0.8'
 # Pin RuboCop for Travis builds.
 gem 'rubocop', '0.54.0'
 
+gem 'pry'
+gem 'pry-byebug'
```
[the remark this answers is not in the document]
Comment (Quintasan) [reply Gemfile]: Done

**Review by sds — APPROVED** [review]

**Comment by sds** [comment]
Thanks!

**Comment by sds** [comment]
@Quintasan I'm seeing a warning once this commit is added. Can you have a look?

```
~/overcommit/lib/overcommit/hook/pre_commit/execute_permissions.rb:29: warning: Overcommit::Hook::PreCommit::Base#initial_commit? at ~/.rbenv/versions
/2.6.5/lib/ruby/2.6.0/forwardable.rb:158 forwarding to private method Overcommit::HookContext::PreCommit#initial_commit?
```

**Comment by sds** [comment]
Thank you for adding this functionality and taking it through to completion, @Quintasan.

Released in [0.56.0][https://github.com/sds/overcommit/releases/tag/v0.56.0).

**Pull request merged**